### ISSUE TITLE
On FreeBSD fftw3 installs lib/libfftw3_mpi.so with unresolved ompi_xx and other symbols

### ISSUE THREAD

**Issue by yurivict** [FFTW/fftw3#169]
It seems to be missing an mpi library in the link line.

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=238008

**Comment by yurivict** [comment]
fftw3 should use ```$(pkg-config --libs ompi-c)``` for ```LDFLAGS``` for ```libfftw3_mpi.so```

**Comment by matteo-frigo** [comment]
I am confused.  The (current) conclusion of the freebsd bugzilla thread seems to be that FFTW is behaving correctly.  What exactly do you need us to do?

**Comment by yurivict** [comment]
You install libfftw3_mpi.so with unresolved symbols ```ompi_xx```.

What you need to do: add ```$(pkg-config --libs ompi-c)``` to the link line for ```lib/libfftw3_mpi.so```.
Otherwise, you install ```lib/libfftw3_mpi.so``` with unresolved OpenMPI symbols.

You should also add ```-Wl,-z,defs``` to ```LDFLAGS```, this would ensure that there are no unresolved symbols. With this flag ```fftw3``` explicitly fails:
```
api.c:(.text+0x8b): undefined reference to `ompi_mpi_comm_null'
/usr/local/bin/ld: api.c:(.text+0x9d): undefined reference to `ompi_mpi_op_sum'
/usr/local/bin/ld: api.c:(.text+0xa6): undefined reference to `ompi_mpi_op_max'
/usr/local/bin/ld: api.c:(.text+0xad): undefined reference to `ompi_mpi_double'
/usr/local/bin/ld: api.c:(.text+0xbf): undefined reference to `MPI_Allreduce'
/usr/local/bin/ld: .libs/api.o: in function `wisdom_ok_hook':
api.c:(.text+0x125): undefined reference to `ompi_mpi_comm_null'
/usr/local/bin/ld: api.c:(.text+0x188): undefined reference to `ompi_mpi_unsigned'
/usr/local/bin/ld: api.c:(.text+0x19c): undefined reference to `MPI_Bcast'
/usr/local/bin/ld: api.c:(.text+0x1d6): undefined reference to `ompi_mpi_int'
/usr/local/bin/ld: api.c:(.text+0x1dd): undefined reference to `ompi_mpi_op_land'
/usr/local/bin/ld: api.c:(.text+0x1f4): undefined reference to `MPI_Allreduce'
/usr/local/bin/ld: api.c:(.text+0x20b): undefined reference to `ompi_mpi_comm_null'
/usr/local/bin/ld: api.c:(.text+0x212): undefined reference to `ompi_mpi_comm_null'
```

**Comment by matteo-frigo** [comment]
@yurivict I don't think we can call "pkg-config --libs ompi-c" unconditionally.  For one thing, if "mpicc" comes from mpich, we would have to call "pkg-config --libs mpich-c".

The MPI code in FFTW is almost 15 years old, and nobody has ever complained about this issue.  

The expectation 15 years ago was that mpi programs would be compiled with "mpicc", and that mpicc adds the appropriate libraries.  Consequently, FFTW uses "mpicc" as a compiler and it does not attempt to divine -I and -L flags.

You seem to imply that this old expectation has changed.  That's fine, and I am not opposed to staying up to date, but we need to do so in a way that does not break anything else.  Your suggestion would break mpich as far as I can tell.  Can you suggest a way to implement your suggestion that works with all (or at least two) MPI implementations?  Do you have an example of some other MPI library that does what you want, so that we can copy what they do?

@stevengj Any suggestions?

**Comment by yurivict** [comment]
you have mpicc, it links executables with the proper mpi library.
When I compiled a simple main function using mpicc it gets linked with the correct mpi library:
```
$ /usr/local/mpi/openmpi/bin/mpicc -o m xmc  && ldd -a m
/usr/ports/math/fftw3/m:
	libmpi.so.12 => /usr/local/mpi/openmpi/lib/libmpi.so.12 (0x800249000)
	libthr.so.3 => /lib/libthr.so.3 (0x80051a000)
	libc.so.7 => /lib/libc.so.7 (0x800547000)
/usr/local/mpi/openmpi/lib/libmpi.so.12:
	libopen-rte.so.12 => /usr/local/mpi/openmpi/lib/libopen-rte.so.12 (0x800940000)
	libopen-pal.so.13 => /usr/local/mpi/openmpi/lib/libopen-pal.so.13 (0x800bb8000)
<...skipped...>
```

So you should just use this way to link the ```libfftw3_mpi.so``` library.

**Comment by matteo-frigo** [comment]
FFTW uses libtool to generate shared libraries.  We pass "mpicc" to libtool, at least on linux:

`/bin/sh ../libtool  --tag=CC   --mode=link mpicc  -O3 -fomit-frame-pointer -mtune=native -malign-double -fstrict-aliasing -fno-schedule-insns -version-info 8:9:5  -o libfftw3_mpi.la -rpath /usr/local/lib any-true.lo api.lo block.lo choose-radix.lo conf.lo dtensor.lo rearrange.lo wisdom-api.lo f03-wrap.lo transpose-alltoall.lo transpose-pairwise.lo transpose-recurse.lo transpose-problem.lo transpose-solve.lo dft-serial.lo dft-rank-geq2.lo dft-rank-geq2-transposed.lo dft-rank1.lo dft-rank1-bigvec.lo dft-problem.lo dft-solve.lo rdft-serial.lo rdft-rank-geq2.lo rdft-rank-geq2-transposed.lo rdft-rank1-bigvec.lo rdft-problem.lo rdft-solve.lo rdft2-serial.lo rdft2-rank-geq2.lo rdft2-rank-geq2-transposed.lo rdft2-problem.lo rdft2-solve.lo ../libfftw3.la   -lm`

However, it appears that libtool still uses gcc for linking:
`libtool: link: gcc -shared  -fPIC -DPIC  .libs/any-true.o .libs/api.o .libs/block.o .libs/choose-radix.o .libs/conf.o .libs/dtensor.o .libs/rearrange.o .libs/wisdom-api.o .libs/f03-wrap.o .libs/transpose-alltoall.o .libs/transpose-pairwise.o .libs/transpose-recurse.o .libs/transpose-problem.o .libs/transpose-solve.o .libs/dft-serial.o .libs/dft-rank-geq2.o .libs/dft-rank-geq2-transposed.o .libs/dft-rank1.o .libs/dft-rank1-bigvec.o .libs/dft-problem.o .libs/dft-solve.o .libs/rdft-serial.o .libs/rdft-rank-geq2.o .libs/rdft-rank-geq2-transposed.o .libs/rdft-rank1-bigvec.o .libs/rdft-problem.o .libs/rdft-solve.o .libs/rdft2-serial.o .libs/rdft2-rank-geq2.o .libs/rdft2-rank-geq2-transposed.o .libs/rdft2-problem.o .libs/rdft2-solve.o   -Wl,-rpath -Wl,/home/athena/sandbox/fftw3/.libs ../.libs/libfftw3.so -lm  -O3 -mtune=native -malign-double   -Wl,-soname -Wl,libfftw3_mpi.so.3 -o .libs/libfftw3_mpi.so.3.5.9`

libtool behaves in this way because `CC=gcc` is hardcoded into the libtool script at configure time.

Given this structure of libtool, there appears to be no convenient way to use gcc to link the main library and mpicc to link the mpi library.  (One workaround would be to use `./configure CC=mpicc`, which would link everything with mpi.)

How big of a problem is this really?  The rest of the world has linked their main program with mpicc for years and things just worked fine.  Is there a reason why you can't link your main program with mpicc?

**Comment by yurivict** [comment]
> How big of a problem is this really? The rest of the world has linked their main program with mpicc for years and things just worked fine. Is there a reason why you can't link your main program with mpicc?

You can call me a perfectionist, but the problem objectively lies in the libfftw3 project itself because it leaves dangling symbols. This isn't a huge problem, generally speaking, but when there are a lot of such problems they collectively become a big problem.

You could pass mpicc to libtool, perhaps through the ```LD``` environment variable. There are other ways too.

**Comment by matteo-frigo** [comment]
> 
> You can call me a perfectionist, but the problem objectively lies in the libfftw3 project itself because it leaves dangling symbols. This isn't a huge problem, generally speaking, but when there are a lot of such problems they collectively become a big problem.

FWIW, static libraries (*.a) have never encoded their dependencies on any system that I know of.  So, the hack of encoding dependencies in shared libraries is just a hack, and it doesn't solve any real problem because one must still be aware of dependencies if one wants to produce static binaries.  For example, when linking a static binary with "-lfftw3", one must add "-lm" to link the sin()/cos() routines.  
So dependency tracking is a half-assed hack in the first place, and the world would be better without it.  Your point would be more convincing if both static and shared libraries behaved in the same way, but that's not the world we live in.

> 
> You could pass mpicc to libtool, perhaps through the `LD` environment variable. There are other ways too.

No you can't.  libtool is a shell scripts that explicitly sets "CC=gcc".  This shell command explicitly overrides any value that CC may have in the environment.  LD is a variable in Makefile, which libtool ignores.  Which other ways are you referring to?

Basically, libtool hardcodes this command at configure time (on linux):
`archive_cmds="\$CC -shared \$pic_flag \$libobjs \$deplibs \$compiler_flags \$wl-soname \$wl\$soname -o \$lib"`

and it hardcodes "CC=gcc" as well.  So there seems to be no way to have a single libtool script that works with both gcc and mpicc.

**Comment by yurivict** [comment]
> No you can't. libtool is a shell script that explicitly sets "CC=gcc". 

This is not true. FreeBSD has CC=cc which is clang, and all libtool-based ports (thousands of them) successfully override the CC=gcc default in libtool.

**Comment by matteo-frigo** [comment]
> This is not true. FreeBSD has CC=cc which is clang, and all libtool-based ports (thousands of them) successfully override the CC=gcc default in libtool.

Right, but you must override CC at configure time, not at link time.  I.e., if you invoke "configure CC=clang", then things work, and they work because configure takes "ltmain.sh", adds a header that says "CC=clang", and calls the combined script "libtool".  So far so good.  But now the problem is, how do you link libfftw3.so with clang and libfftw3_mpi.so with mpicc?  You can't because CC=clang is hardcoded in the generated libtool script.

The real problem is that "mpicc" is a bad idea.  It is insane that the official way to use a certain library is to use a special compiler.  Imagine if we insisted that you must use fftwcc in order to use fftw.  But MPI insists that the only official way to link with the MPI library is to use mpicc, and I am afraid that your best bet is to submit to this nonsense and move on.

**Comment by yurivict** [comment]
This is not resolved. Why did you close it? The problem is still there.

GNU tools suggested a solution to read the libraries from ```mpicc``` but I didn't have time to re-type it here yet.

**Comment by matteo-frigo** [comment]
I closed this because it's not a bug.  MPI has decided that MPI programs must be linked with mpicc.  There is nothing I can do about this.  Even if you come up with a hack that works with .so files, the hack won't work with .a files anyway.

**Comment by yurivict** [comment]
.a files aren't affected by the problem since their dependencies are supplied at a very top level anyway.
